### PR TITLE
Checkpoints

### PR DESCRIPTION
**Pull request by GainiK**
checkpoints callbacks:
 - logs folder in package's root folder
 - name is as wandb run id

### CODE REVIEW

File: seg_lapa/config/trainer/trainer.yaml [hunk 1]
```diff
@@ -8,7 +8,7 @@ precision: 16
 limit_train_batches: 1.0
 limit_val_batches: 1.0
 limit_test_batches: 1.0
-fast_dev_run: False
+fast_dev_run: True
```
Comment (Shreeyak) [inline seg_lapa/config/trainer/trainer.yaml]: Please reset this to false

File: seg_lapa/config_parse/checkpoint_conf.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        pass
+
+
+@dataclass
+class CheckpointConfig(CheckpointConf):
+    dirpath: Optional[str]
+
+    def get_checkpoint_callback(self, cfg: DictConfig, run_id) -> pl.callbacks:
+        if self.dirpath is None:
+            self.dirpath = str(get_project_root())+'/logs/'+run_id
```
Comment (Shreeyak) [inline seg_lapa/config_parse/checkpoint_conf.py]: Use pathlib library to deal with paths

File: seg_lapa/config_parse/checkpoint_conf.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        checkpoint_callback = ModelCheckpoint(**asdict_filtered(self))
```
Comment (Shreeyak) [inline seg_lapa/config_parse/checkpoint_conf.py]: This saves the checkpoints for a run in the given directory, yes?

What other logs are generated by lightning? The trainer has a root_dir parameter. That should also be set
That param will control where the pytorch-lightning-logs dir is created. But will probably still create subdirs called  version0, version1, etc. We'd ideally like the change the naming scheme of those subdirs to use wandb runid instead and save the checkpoints there as well.

File: seg_lapa/config/train.yaml [hunk 4]
```diff
@@ -5,6 +5,7 @@ defaults:
   - trainer: trainer
   - scheduler: disabled
   - logger: wandb
+  - checkpoint_callback: checkpoint_callback
```
Comment (Shreeyak) [inline seg_lapa/config/train.yaml]: Please restructure the code to follow format in early stopping PR. I've created methods to retrieve a list of callbacks. Add this model checkpoint callback to the callbacks/standard.yaml file.

File: seg_lapa/config_parse/train_conf.py [hunk 5]
```diff
@@ -18,6 +19,7 @@ class TrainConf:
     trainer: TrainerConf
     scheduler: SchedulerConf
     logger: LoggerConf
+    checkpoint_callback: CheckpointConf
```
Comment (Shreeyak) [inline seg_lapa/config_parse/train_conf.py]: Same as before. Please restructure code to follow format in early stopping PR.

**Review by Shreeyak — CHANGES_REQUESTED** [review]
Needs restructuring of the config system, per Early Stopping PR.

File: seg_lapa/config_parse/checkpoint_conf.py (hunk 3 above)
Comment (GainiK) [reply seg_lapa/config_parse/checkpoint_conf.py]: This will create the run_id folder inside logs folder, and save checkpoints inside the run_id folder. This is expected behavior, right?
Will check what other logs are generated by lightning. 
Wandb logs are also generated. Should I disable it?

Comment (Shreeyak) [reply seg_lapa/config_parse/checkpoint_conf.py]: Yes, I understand. My point is that pl itself creates a logs folder. Using their format,
It's easy to retrieve the location of current log folder and save other stuff locally as required (images, point clouds, etc). So I'd prefer to modify pl's system of creating log directories.
The root log for can be set with param to trainer.

Comment (Shreeyak) [reply seg_lapa/config_parse/checkpoint_conf.py]: Wandb logs - what all do they generate? Not sure it even can be disabled. I certainly never use it.

File: seg_lapa/config_parse/callbacks_conf.py [hunk 6]
```diff
@@ -16,8 +17,8 @@ class CallbacksConf(ABC):
     def get_callbacks_dict(self):
         return {}
 
-    def get_callbacks_list(self):
-        callback_dict = self.get_callbacks_dict()
+    def get_callbacks_list(self, log_dir):
+        callback_dict = self.get_callbacks_dict(log_dir)
```
Comment (Shreeyak) [inline seg_lapa/config_parse/callbacks_conf.py]:
```suggestion
    def get_callbacks_list(self, *args):
        callback_dict = self.get_callbacks_dict(*args)
```
Since this is an abstract class, it's better to indicate that some optional positional arguments may be passed by `*args`. For named arguments, we'd use `**kwargs`.

This way, if the child classes want, they may accept additional positional arguments, but are not forced to.

File: seg_lapa/train.py [hunk 7]
```diff
@@ -11,14 +11,16 @@
...
 
 
 class DeeplabV3plus(pl.LightningModule):
-    def __init__(self, config: TrainConf):
+    def __init__(self, config: TrainConf, run_id: str):
         super().__init__()
         self.cross_entropy_loss = CrossEntropy2D(loss_per_image=True, ignore_index=255)
         self.config = config
         self.model = self.config.model.get_model()
+        self.logs_dir = get_project_root() / 'logs' / run_id
```
Comment (Shreeyak) [inline seg_lapa/train.py]: 1. If the directory does not already exist, we need to create it. 
2. Any constants such as `'logs'` should be saved as a constant somewhere (top of file or init of class)

Let's create a function to create logs dir and directly pass the path to logs dir to this class

File: seg_lapa/train.py [hunk 8]
```diff
@@ -114,9 +116,9 @@ def main(cfg: DictConfig):
         print("\nResolved Dataclass:\n", config, "\n")
 
     wb_logger = config.logger.get_logger(cfg)
-    callbacks = config.callbacks.get_callbacks_list()
+    model = DeeplabV3plus(config, wb_logger.experiment.id)
+    callbacks = config.callbacks.get_callbacks_list(model.logs_dir)
```
Comment (Shreeyak) [inline seg_lapa/train.py]: It feels weird to couple callbacks and LightningModule like this. 
Let's create a function in train.py that creates the log directory, and pass this directory to both the pl.LightningModule and the Callbacks:

We can also generate a run_id and use that to init wandb
```
    LOGS_DIR = 'logs'

    def create_log_dir(self, cfg):
        # Set the run ID: Read from config if resuming training, else generate unique id
        run_id = wandb.util.generate_id()
        resume_run = "never"  # Alternate: "allow"

        # Create log dir
        log_root_dir = get_project_root() / LOGS_DIR
        log_dir = log_root_dir / run_id
        log_dir.mkdir(parents=True, exist_ok=True)

        # Save the input config file to logs dir
        OmegaConf.save(cfg, log_dir / "train.yaml")
        
        return log_dir, run_id
```

You'll need to modify the WandbLogger:
1. Remove run_id from wandb config. Init wandb using our generated run id.
2. Set the save_dir of wandb to also be in project root

```
wb_logger = config.logger.get_logger(cfg, run_id)
```
```
@dataclass(frozen=True)
class WandbConf(LoggerConf):
    """Weights and Biases. Ref: wandb.com"""

    entity: str
    project: str
    run_name: Optional[str]

def get_logger(self, cfg: DictConfig, run_id: str, save_dir: str) -> pl_loggers.WandbLogger:
    wb_logger = pl_loggers.WandbLogger(name=run_name, id=run_id, save_dir=save_dir config=cfg_dict, **args_dict)

```

**Review by Shreeyak — CHANGES_REQUESTED** [review]

File: seg_lapa/train.py (hunk 7 above)
Comment (GainiK) [reply seg_lapa/train.py]: 1. Lightning takes care about creating directory if it does not exist

Comment (Shreeyak) [reply seg_lapa/train.py]: Huh? Could you please link to the docs page or code sample. 
This is literally an attribute we're adding to the lightning module, how does it create the dir?

File: seg_lapa/config_parse/callbacks_conf.py [hunk 9]
```diff
@@ -40,15 +41,25 @@ def get_callback(self):
         return EarlyStopping(**args_dict)
 
 
+@dataclass(frozen=True)
+class CheckpointConf:
+    """Dataclass just to initialize and return the Checkpoint Callback"""
+
+    def get_callback(self, logs_dir):
```
Comment (Shreeyak) [inline seg_lapa/config_parse/callbacks_conf.py]:
```suggestion
class CheckpointConf:
    """Dataclass just to initialize and return the Checkpoint Callback"""
    monitor: Optional[str]
    mode: str
    save_last: Optional[bool]
    period: int
    

    def get_callback(self, logs_dir):
```
Please add the other arguments to Modelcheckpoint to the .yaml config file and to this Conf class. These are important parameters to be able to control. 
In our case, we want to monitor `Val/loss` and mode should be `max`.

https://pytorch-lightning.readthedocs.io/en/latest/generated/pytorch_lightning.callbacks.ModelCheckpoint.html#pytorch_lightning.callbacks.ModelCheckpoint

File: seg_lapa/config_parse/callbacks_conf.py [hunk 10]
```diff
@@ -40,15 +41,25 @@ def get_callback(self):
...
+        checkpoint_callback = ModelCheckpoint(dirpath=logs_dir)
+        return checkpoint_callback
+
+
 @dataclass(frozen=True)
 class StandardCallbacksConf(CallbacksConf):
     """Get a dictionary of all the callbacks."""
 
     early_stopping: Dict
```
Comment (Shreeyak) [inline seg_lapa/config_parse/callbacks_conf.py:64]: Once we add the model checkpoints vales to .yaml file, we'll need to add an entry here too.

File: seg_lapa/config_parse/logger_conf.py [hunk 11]
```diff
@@ -54,6 +54,8 @@ def get_logger(self, cfg: DictConfig) -> pl_loggers.WandbLogger:
         # The argument names to wandb are different from the attribute names of the class.
         # Pop the offending attributes before passing to init func.
         args_dict = asdict_filtered(self)
+        if args_dict["save_dir"] is None:
+            args_dict["save_dir"] = save_dir
         run_name = args_dict.pop("run_name")
         run_id = args_dict.pop("run_id")
```
Comment (Shreeyak) [inline seg_lapa/config_parse/logger_conf.py]:
```suggestion
```
We're overwriting the run_id argument.

At this point, I don't think we need the run_id in config file. Please remove corresponding entry in .yaml config files and it's dataclass.

**Review by Shreeyak — CHANGES_REQUESTED** [review]
Looking good, almost there!

File: seg_lapa/train.py (hunk 7 above)
Comment (Shreeyak) [reply seg_lapa/train.py]: @GainiK Do you mean by setting `self.log_dir` in LightningModule, pl will create the dir for us?

Comment (Shreeyak) [reply seg_lapa/train.py]: (this is different from the `self.logs_dir` above.)

File: seg_lapa/config/callbacks/standard.yaml [hunk 12]
```diff
@@ -3,3 +3,8 @@ name: standard
 early_stopping:
   min_delta: 0.0005
   patience: 10
+checkpoints:
+  monitor: 'val_loss'
```
Comment (Shreeyak) [inline seg_lapa/config/callbacks/standard.yaml]: Is there any way to monitor the IoU instead? We do have self.log(Val/miou)

File: seg_lapa/config/callbacks/standard.yaml [hunk 13]
```diff
@@ -3,3 +3,8 @@ name: standard
...
+  mode: 'max'
```
Comment (Shreeyak) [inline seg_lapa/config/callbacks/standard.yaml:8]: For loss, this should be min. For IoU, it'll be max.

**Review by Shreeyak — CHANGES_REQUESTED** [review]
Awesome! One last change - let's monitor IoU instead of loss.
Also, do verify it works by running for a few epochs.

**Review by Shreeyak — COMMENTED** [review]
LGTM!

**Review by Shreeyak — APPROVED** [review]
